db/repair_bucket.go

```diff
@@ -75,82 +76,170 @@ func (r *RepairBucket) InitFrom(params RepairBucketParams) *RepairBucket {
+                                                        Limit: 3
+                                                    NumProcessed: 0
+
+* It starts with an empty start key
+* For the next page, it uses the last key processed as the new start key
+* Since the start key is inclusive, it will see the start key twice (on first page, and on next page)
+* If it's iterating a result page and sees a doc with the start key (eg, doc3 in above), it will ignore it so it doesn't process it twice
+* Stop condition: if NumProcessed is 0, because the only doc in result set had already been processed.
+*
+*/
 func (r RepairBucket) RepairBucket() (results []RepairBucketResult, err error) {
 
+	base.LogTo("CRUD", "RepairBucket() invoked")
+	defer base.LogTo("CRUD", "RepairBucket() finished")
+
+	startKey := ""
 	results = []RepairBucketResult{}
-	options := Body{"stale": false, "reduce": false}
-	options["startkey"] = []interface{}{true}
 
-	vres, err := r.Bucket.View(DesignDocSyncHousekeeping, ViewImport, options)
-	if err != nil {
-		return results, err
-	}
+	for {
 
-	for _, row := range vres.Rows {
-		rowKey := row.Key.([]interface{})
-		docid := rowKey[1].(string)
-		key := realDocID(docid)
-		var backupOrDryRunDocId string
+		options := Body{"stale": false, "reduce": false}
+		options["startkey"] = []interface{}{
+			true,
+			startKey,
+		}
+		options["limit"] = base.ViewQueryPageSize
 
-		err = r.Bucket.Update(key, 0, func(currentValue []byte) ([]byte, error) {
-			// Be careful: this block can be invoked multiple times if there are races!
-			if currentValue == nil {
-				return nil, couchbase.UpdateCancel // someone deleted it?!
-			}
+		base.LogTo("CRUD", "RepairBucket() querying view with options: %+v", options)
+		vres, err := r.Bucket.View(DesignDocSyncHousekeeping, ViewImport, options)
+		base.LogTo("CRUD", "RepairBucket() queried view and got %d results / total rows: %d", len(vres.Rows), vres.TotalRows)
+		if err != nil {
+			return results, err
+		}
 
-			updatedDoc, shouldUpdate, repairJobs, err := r.TransformBucketDoc(key, currentValue)
-			if err != nil {
-				return nil, err
+		// Check
+		if len(vres.Rows) == 0 {
+			// No more results.  Return
+			return results, nil
+		}
+
+		// Keep a counter of how many results were processed, since if none were processed it indicates that
+		// we hit the last (empty) page of data.  This is needed because the start key is inclusive, and
+		// so even on the last page of results, the view query will return a single result with the start key doc.
+		numResultsProcessed := 0
+
+		for _, row := range vres.Rows {
+
+			rowKey := row.Key.([]interface{})
+			docid := rowKey[1].(string)
+
+			if docid == startKey {
+				// Skip this, already processed in previous iteration.  Important to do this before numResultsProcessed
+				// is incremented.
+				continue
 			}
 
-			switch shouldUpdate {
-			case true:
+			log.Printf("process doc id: %v", docid)
```

Comment:
Someone important told me this should be removed ;)

Reply:
Fixed